typescript/infra/src/utils/utils.ts

```diff
@@ -240,3 +240,39 @@ export function streamToString(stream: NodeJS.ReadableStream): Promise<string> {
       .on('end', () => resolve(String.prototype.concat(...chunks)));
   });
 }
+
+/**
+ * Converts a matrix to 1d array ordered by diagonals. This is useful if you
+ * want to make sure that the order operations are performed in are ordered but
+ * not repeating the same values from the inner or outer array in sequence.
+ *
+ * @warn Requires a square matrix.
+ *
+ * // 0,0 1,0 2,0 3,0
+ * //
+ * // 0,1 1,1 2,1 3,1
+ * //
+ * // 0,2 1,2 2,2 3,2
+ * //
+ * // 0,3 1,3 2,3 3,3
+ *
+ * becomes
+ *
+ * 0,0; 1,0; 0,1; 2,0; 1,1; 0,2; 3,0; 2,1; 1,2; 0,3; 3,1; 2,2; 1,3; 3,2; 2,3; 3,3
+ *
+ * Adapted from
+ * https://www.geeksforgeeks.org/zigzag-or-diagonal-traversal-of-matrix/
+ */
+export function diagonalize<T>(array: Array<Array<T>>): Array<T> {
```

Comment:
I find this pretty confusing but don't have a better suggestion

Reply:
This was painful. I thought "Oh this will be easy" and pretty quickly regretted that so I just tried to document well and copy paste and test.